nomad/state/state_store.go

```diff
@@ -3177,6 +3177,55 @@ func (s *StateStore) EvalByID(ws memdb.WatchSet, id string) (*structs.Evaluation
 	return nil, nil
 }
 
+// EvalsRelatedToID is used to retrieve the evals that are related (next,
+// previous, or blocked) to the provided eval ID.
+func (s *StateStore) EvalsRelatedToID(ws memdb.WatchSet, id string) ([]*structs.EvaluationStub, error) {
+	txn := s.db.ReadTxn()
+
+	raw, err := txn.First("evals", "id", id)
+	if err != nil {
+		return nil, fmt.Errorf("eval lookup failed: %v", err)
+	}
+	if raw == nil {
+		return nil, nil
+	}
+	eval := raw.(*structs.Evaluation)
+
+	relatedEvals := []*structs.EvaluationStub{}
+	todo := eval.RelatedIDs()
+	done := map[string]bool{
+		eval.ID: true, // don't place the requested eval in the related list.
+	}
+
+	for len(todo) > 0 {
+		// Pop the first value from the todo list.
+		current := todo[0]
+		todo = todo[1:]
+		if current == "" {
+			continue
+		}
+
+		// Skip value if we already have it in the results.
+		if done[current] {
+			continue
+		}
+
+		eval, err := s.EvalByID(ws, current)
+		if err != nil {
+			return nil, err
+		}
+		if eval == nil {
+			continue
+		}
+
+		todo = append(todo, eval.RelatedIDs()...)
+		relatedEvals = append(relatedEvals, eval.Stub())
+		done[eval.ID] = true
+	}
```

Comment:
Currently the only way out of this loop is to scan all related evals (or experience an error). Is there a way to limit the scanning we do, like on the size of the accumulated `relatedEvals` or even just a hard stop at like 20?

Reply:
Hum...good question. I think the problem with setting an upper bound is that the response then becomes incomplete, and the idea of this flag is to give all the data needed quickly. But guarding against potential infinite loops, or slow requests sounds like a good idea. I don't know what a good number would be, so perhaps we can have a time limit, like 1s?

Reply:
Maybe the most correct thing to do would be to plumb the `Request.Context` all the way down here. (You could make the same argument for all calls into the state store, it's just that this one feels less strongly bounded than the others)

Reply:
Ah good point. Is this the channel I should be listening for connection close? https://pkg.go.dev/github.com/hashicorp/yamux#Session.CloseChan

Reply:
I think we want the http request context

https://pkg.go.dev/net/http#Request.Context

stemming from the request handler http.Request

https://github.com/hashicorp/nomad/blob/main/command/agent/eval_endpoint.go#L10

Reply:
Ah ok. Unfortunately that's currently not possible 😬 
https://github.com/hashicorp/nomad/issues/7119

Reply:
Whelp so much for that! It's probably fine as-is anyway, I mean how many related evals could there be anyway ... :thinking:

Reply:
fwiw, another endpoint returns all the evaluations for a job in one go, so this is strictly less than that... can't be too bad, right?

Reply:
Ah yeah, good point. One of the differences here is that we have a (potentially) infinite loop if the `todo` list never gets to zero. But I think we're safe because we always pop and check if eval has been processed already.